### PR TITLE
ARROW-2141: [Python] Support variable length binary conversion from Pandas

### PR DESCRIPTION
**Pull request by BryanCutler**
Currently, when performing `from_pandas` conversion with binary data and the user specifies the type as variable length binary `pa.binary()` then the type is inferred and a cast from binary to binary is attempted.  The casting then fails because the cast kernel does not support binary types.  This PR checks if the user specifies a variable length binary type in conversion, and then copies data to a BinaryArray instead of trying to infer the type and then casting.

### CODE REVIEW

**Comment by BryanCutler** [comment]
@wesm and @xhochy , hopefully this is the right approach.  Do you guys think that conversion from `bytearray`s could be supported too?  That is what pyspark requires binary data to be for some reason..

File: cpp/src/arrow/python/numpy_to_arrow.cc
Comment (pitrou) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: This line looks misindented. Does your editor insert tabs instead of spaces?

**Comment by pitrou** [comment]
See also linting failure at https://travis-ci.org/apache/arrow/jobs/348494207#L750

**Comment by pitrou** [comment]
> Do you guys think that conversion from bytearrays could be supported too? That is what pyspark requires binary data to be for some reason..

That sounds reasonable to me. Actually, any object supporting the buffer protocol should do.

File: cpp/src/arrow/python/numpy_to_arrow.cc
Comment (BryanCutler) [reply cpp/src/arrow/python/numpy_to_arrow.cc]: oops, let me fix that.

**Comment by BryanCutler** [comment]
Thanks @pitrou!  Let me fix the formatting issues and I'll look into supporting bytearrays.

File: cpp/src/arrow/python/numpy_to_arrow.cc
Comment (pitrou) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: We're probably doing this kind of dance (taking a bytes object and extracting a pointer and size) in other places already. I think it would be nice to factor that that somewhere (see e.g. `src/arrow/python/helpers.h`). That would also allow supporting bytearray in other places.

File: cpp/src/arrow/python/numpy_to_arrow.cc
Comment (pitrou) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: I have trouble parsing this... If we don't know how to convert this, we should fail, not simply break from the loop, no?

Comment (pitrou) [reply cpp/src/arrow/python/numpy_to_arrow.cc]: After reading the code a bit more carefully, I understand... though there is still a problem: what if `length` is larger than `kBinaryMemoryLimit`?

File: cpp/src/arrow/python/numpy_to_arrow.cc
Comment (pitrou) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: The cast is entirely wrong :-(

**Comment by wesm** [comment]
I moved this JIRA to 0.10.0 so we can give this situation a working over

**Comment by BryanCutler** [comment]
No problem to hold off on this for a while, seems like there are maybe some issues that @pitrou pointed out that need a deeper look

**Comment by BryanCutler** [comment]
@wesm and @pitrou, since the issues brought up here did not originate from this pr, do you think this could be merged and then follow up with another JIRA to look at these issues?  This is blocking https://issues.apache.org/jira/browse/SPARK-23555 so if possible I'd like to get it resolved.

**Comment by pitrou** [comment]
Can you open the relevant JIRA issue?

Also, before this PR is merged, the Travis-CI failures should be fixed.

**Comment by BryanCutler** [comment]
I made https://issues.apache.org/jira/browse/ARROW-2380 to cover the issues brought up

**Comment by BryanCutler** [comment]
The remaining Travis failures are with GLib builds and seem unrelated

**Review by xhochy — APPROVED** [review]
+1, LGTM

I'm letting @pitrou merge this once he also made a final review.

**Review by pitrou — APPROVED** [review]
+1. @xhochy, I don't have merge rights yet, so it may be quicker for you to merge it.

**Comment by BryanCutler** [comment]
Merged to master, thanks @pitrou and @xhochy!

**Pull request closed without merging**
